pkg/util/docker/dockerfile/instructions.go

```diff
@@ -0,0 +1,68 @@
+package dockerfile
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/docker/docker/builder/command"
+)
+
+// A KeyValue can be used to build ordered lists of key-value pairs.
+type KeyValue struct {
```

Comment:
The three types are conceptually different, and we need to keep the dependencies clear.
WRT `kapi.EnvVar` and downward API, we'd process references to build fields before calling any function from the `pkg/util/docker/dockerfile` package.
This util package should know only about Dockerfiles, not about Kubernetes nor downwards API (as it would then have to know about builds as well, and we'd be making a hell of a mess there)...
That's why we convert from `kapi.EnvVar` to `dockerfile.KeyValue`, just like before we were already converting `kapi.EnvVar` into `map[string]string` used by S2I -- and, as I noted with a [comment in the code](https://github.com/openshift/origin/pull/4562/files#diff-b9627ec6ab0689e50e5fa3e5d958c1d3R164), using a map is a bad idea. To fix that we'll need to revisit S2I.